### PR TITLE
Fix incorrect test set values in leave_k_out splits with sparse user rows

### LINKED ISSUE #639: leave_k_out_split produces incorrect values in test set
I noticed odd behavior in [leave_k_out_split](https://github.com/benfred/implicit/blob/871e0c7229b012108131b6211cd617e23a3b24bf/implicit/evaluation.pyx#L143) - specifically, under certain circumstances (many rows with k < N), the number returned for the withheld test set value is a multiple of the actual value. This causes `train + test` to fail to reconstruct the original matrix.

```python
from implicit.evaluation import leave_k_out_split
from scipy import sparse
import numpy as np

rng = np.random.RandomState(42)

ratings = sparse.csr_matrix(
    [[3, 2, 1, 0], 
     [1, 0, 0, 0], 
     [0, 1, 0, 0], 
     [0, 0, 1, 0], 
     [0, 1, 1, 1]]
)

train, test = leave_k_out_split(ratings, K=1, random_state=rng)

diff = (train + test) - ratings

diff.nnz
#> 1

train.todense()
#> matrix([[0, 2, 1, 0],
#>         [1, 0, 0, 0],
#>         [0, 1, 0, 0],
#>         [0, 0, 1, 0],
#>         [0, 0, 1, 1]])

test.todense()
#> matrix([[12,  0,  0,  0],
#>         [ 0,  0,  0,  0],
#>         [ 0,  0,  0,  0],
#>         [ 0,  0,  0,  0],
#>         [ 0,  1,  0,  0]])

diff.todense()
#> matrix([[9, 0, 0, 0],
#>         [0, 0, 0, 0],
#>         [0, 0, 0, 0],
#>         [0, 0, 0, 0],
#>         [0, 0, 0, 0]])
```

<sup>Created at 2022-12-23 13:19:25 CST by [reprexlite](https://github.com/jayqi/reprexlite) v0.5.0</sup>

In addition, the call to `_take_tails` when `shuffled=True` does not pass on the `rng`, so the random state cannot be maintained.

### PR DESCRIPTION
**Pull request by chrisjkuch**
Closes #639 

This PR fixes a bug in the evaluation of the `leave_k_out_split` in which the produced test matrix would contain values that were many multiples of their original value. Tests are also added on static (non-random) matrices that otherwise fail in the un-corrected implementation.

This bug resulted from a calculation that required an input array with sequential values - the fact that non-sequential values were provided led to an error in processing.

Specifically, the `arr` argument in [_take_tails](https://github.com/benfred/implicit/blob/871e0c7229b012108131b6211cd617e23a3b24bf/implicit/evaluation.pyx#L79)

>  
    ----------
    arr: ndarray
        The input array. This should be an array of integers in the range 0->n, where
        the ordered unique set of integers in said array should produce an array of
        consecutive integers. Concretely, the array [1, 0, 1, 1, 0, 3] would be invalid,
        but the array [1, 0, 1, 1, 0, 2] would not be.

was being provided as `candidate_users`, from which user indices falling below the threshold were removed, resulting in a list in which the ordered set of unique integers was not consecutive and therefore the provided array was invalid.

### CODE REVIEW

**Comment by chrisjkuch** [comment]
It looks like a single test failed in one of the builds:

```
=================================== FAILURES ===================================
________________ test_leave_k_out_gets_correct_train_only_shape ________________

    def test_leave_k_out_gets_correct_train_only_shape():
        """Test that the correct number of users appear *only* in the train set."""
    
        mat = _get_matrix()
        train, test = leave_k_out_split(mat, K=1, train_only_size=0.8)
        train_only = ~np.isin(np.unique(train.tocoo().row), test.tocoo().row)
    
>       assert train_only.sum() == int(train.shape[0] * 0.8)
E       assert 81 == 80
```

Given that the tests are only failing in a single build and passing in all others, my guess is that a completely null row was present in the randomly generated matrix that caused it to be included in the training set in addition to all other randomly chosen users. I think the best solution to this is to add a check onto _get_matrix() that ensures that there aren't any completely zero rows.

**Comment by chrisjkuch** [comment]
Hmmm, even after making the fix to ensure always-populated rows, the test is still failing intermittently, and it's failing intermittently both for the random 100x100 sparse matrix as well as the newly-added fixed matrix. 

Usually, the value for the number of users is only slightly off of the chosen value. @benfred would an `almostEquals` here with a `delta` of ~5 or so suffice?

**Comment by chrisjkuch** [comment]
@benfred Looks like this last failing test is addressed by #652, and some of the fixes in this PR are in effect duplicated by #653. Let me know if / how you'd like to proceed in fixing the functionality of `leave_k_out_split`, happy to help in any way I can.
